### ISSUE TITLE
RNNTrainer - AR and TAR broken

### ISSUE THREAD

**Issue by tpietruszka** [fastai/fastai#1460]
**Describe the bug**
In the current version I see a few problems with the [RNNTrainer](https://github.com/fastai/fastai/blob/af4773bbba8fa535d8e9e0af0879a12acc174046/fastai/callbacks/rnn.py): 1 big, 1 small, 1 potential.

1. The big problem. I believe the TAR term (line 31):
```last_loss += (self.beta * (h[1:] - h[:-1]).pow(2).mean()).sum().float()```
is calculated along the wrong axis. Looking at dimensions of `h[1:] - h[:-1]`, we see 1 dimension that is 1 smaller than `batch_size`. The activation differences should be calculated between time steps. Looks like they are being calculated between sequences.

2. The small problem. The AR term (line 28):
``` last_loss += (self.alpha * self.out[-1].pow(2).mean()).sum().float() ```
mean() calculates the mean of all values, returning a single value (rank 0 tensor). Then sum() does nothing. According to the paper, I believe the mean() should return a rank-1 tensor with len = `batch_size`. Alternatively, we could take the result of mean() as is and multiply by `batch_size`.

3. The probably-no-problem:
in line 26, learn.opt.lr is being multiplied and replaced by x, ` 0 < x <= 1`. Is it being restored to the previous value at some point, before the next batch begins?
Otherwise it could decrease the lr because of shorter sequences in the current batch, and then leave it. Lets remember, that the shorter sequences are not necessarily just in the last batch of the epoch, like in `LanguageModel`s - this is also used by classifiers etc.

BTW, @sgugger , I am working on some code that would additionally handle RNN outputs in `PackedSequence` format, and change `fastai/text/models.py` to use that format. 

Would you accept such a change? Should I expect signifficant API changes in this part of the codebase?

Similar work has been discussed and mostly done in https://forums.fast.ai/t/use-packedsequence-in-rnncore-to-improve-pooling-accuracy-and-rnn-speed/27263/16 but it looks like it was never completed and merged.

While I do not think it will speed things up, I think it is still important for 2 reasons:

1. In training, we would expect the network to learn to do nothing when seeing the PAD token. However, it does not really learn that, and the training is harder. I would expect minor accuracy improvements when we get rid of padding.
2. At inference time, results depend on the amount of padding added. I was quite surprised to see that this impact can be significant. In one case, binary classification, I compared the results of running the model on samples scored in batches vs one-by-one. Around 1% of classification decisions were filpped! Also, this is making models de-facto non-deterministic, which I think should be avoided.

**Provide your installation details**
=== Software === 
python        : 3.6.8
fastai        : 1.0.40.dev0
fastprogress  : 0.1.18
torch         : 1.0.0
nvidia driver : 410.78
torch cuda    : 9.0.176 / is available
torch cudnn   : 7401 / is enabled

=== Hardware === 
nvidia gpus   : 1
torch devices : 1
  - gpu0      : 8116MB | GeForce GTX 1080

=== Environment === 
platform      : Linux-4.15.0-43-generic-x86_64-with-debian-buster-sid
distro        : Ubuntu 18.04 Bionic Beaver
conda env     : fastai-dev
python        : /home/tomasz/anaconda3/envs/fastai-dev/bin/python
sys.path      : 
/home/tomasz/anaconda3/envs/fastai-dev/lib/python36.zip
/home/tomasz/anaconda3/envs/fastai-dev/lib/python3.6
/home/tomasz/anaconda3/envs/fastai-dev/lib/python3.6/lib-dynload
/home/tomasz/.local/lib/python3.6/site-packages
/home/tomasz/anaconda3/envs/fastai-dev/lib/python3.6/site-packages
/home/tomasz/anaconda3/envs/fastai-dev/lib/python3.6/site-packages/wikiextractor-2.69-py3.6.egg
/data/users/tomasz/fastai
/home/tomasz/anaconda3/envs/fastai-dev/lib/python3.6/site-packages/IPython/extensions
/home/tomasz/.ipython

**To Reproduce**
```
from fastai.text import *
from fastai.callbacks.rnn import *
path = untar_data(URLs.IMDB_SAMPLE)
df = pd.read_csv(path/'texts.csv', header=None)
data_lm = TextLMDataBunch.from_csv(path, 'texts.csv', bs=42, bptt=75)

def on_backward_begin(self, last_loss:Rank0Tensor, last_input:Tensor, **kwargs):
    "Adjusts the lr to the sequence length and applies AR and TAR to `last_loss`."
    if self.adjust: self.learn.opt.lr *= last_input.size(1) / self.bptt
    #AR and TAR
    if self.alpha != 0.:  last_loss += (self.alpha * self.out[-1].pow(2).mean()).sum().float()
    print(f"AR term before sum(): {self.out[-1].pow(2).mean()}")
    if self.beta != 0.:
        h = self.raw_out[-1]
        if len(h)>1: last_loss += (self.beta * (h[1:] - h[:-1]).pow(2).mean()).sum().float()
    print(f"TAR: (h[1:] - h[:-1]).shape: {(h[1:] - h[:-1]).shape}")
    return last_loss

RNNTrainer.on_backward_begin = on_backward_begin



moms = (0.8,0.7)
learn = language_model_learner(data_lm, pretrained_model=URLs.WT103_1)
learn.unfreeze()
learn.fit_one_cycle(1, slice(1e-2), moms=moms)
```
**Expected behavior**
TAR: (h[1:] - h[:-1]).shape: torch.Size([**42, 74**, 400])
AR term before sum() should have size  [42]

Ideally, I think computing those terms should be moved to separate functions, to be testable. Perhaps also assertions on tensor dimensions?

**Output (beginning of the)**
```text
AR term before sum(): 0.03718101978302002
TAR: (h[1:] - h[:-1]).shape: torch.Size([41, 75, 400])
```

**Comment by sgugger** [comment]
Fixed 1, I forgot to adapt when we moved to batch first. For 2, the paper mentions the L2 norm (which is a mean) and the authors' implementation uses the mean too: https://github.com/salesforce/awd-lstm-lm/blob/32fcb42562aeb5c7e6c9dec3f2a3baaaf68a5cb5/main.py#L201 so I'm leaving it as is.
For 3, I removed the adjust parameter since we're now using a constant bptt (we shuffle the texts at each epoch so it doesn't help anymore).

The correct behavior with padding would be a great addition if it's not slower (there has been an attempt, but it was really slower).
